src/Workspaces/Core/Desktop/Utilities/RemoteTask.cs

```diff
@@ -0,0 +1,120 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface IRemoteTask<T>
+    {
+        void Cancel();
+        void SetCompletion(IRemoteTaskCompletion<T> completion);
+        T GetResult();
+    }
+
+    internal interface IRemoteTaskCompletion<T>
+    {
+        void Complete(T result);
+        void Fail(Exception exception);
+    }
+
+    internal static class RemoteTask
+    {
+        public static Task<TResult> ToTask<TResult>(this IRemoteTask<TResult> asyncTask, CancellationToken cancellationToken)
+        {
+            return new RemoteTaskCompletion<TResult>(asyncTask, cancellationToken).Task;
+        }
+    }
+
+    internal class RemoteTaskCompletion<T> : MarshalByRefObject, IRemoteTaskCompletion<T>
+    {
+        private readonly TaskCompletionSource<T> _taskSource;
+
+        public RemoteTaskCompletion(IRemoteTask<T> asyncTask, CancellationToken cancellationToken)
+        {
+            _taskSource = new TaskCompletionSource<T>();
+
+            asyncTask.SetCompletion(this);
+
+            if (cancellationToken.CanBeCanceled)
+            {
+                cancellationToken.Register(() =>
+                {
+                    asyncTask.Cancel();
+                    _taskSource.TrySetCanceled();
+                });
+            }
+        }
+
+        public Task<T> Task
+        {
+            get { return _taskSource.Task; }
+        }
+
+        void IRemoteTaskCompletion<T>.Complete(T result)
+        {
+            _taskSource.TrySetResult(result);
+        }
+
+        void IRemoteTaskCompletion<T>.Fail(Exception exception)
+        {
+            _taskSource.TrySetException(exception);
+        }
+    }
+
+    internal class RemoteTask<TResult> : MarshalByRefObject, IRemoteTask<TResult>
```

Comment:
Given I just got bitten by remote memory issues for another internal bug due to `MarshalByRefObject`, have we given some thought about lifetime tracking here, especially given issue #3078?